std/experimental/ndslice/algorithm.d

```diff
@@ -0,0 +1,1535 @@
+can be used as they accept a set of tensors instead of single one.
+
+Note:
+    $(SUBREF iteration, transposed) and
+    $(SUBREF selection, pack) can be used to specify dimensions.
+
+License:   $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors:   Ilya Yaroshenko
+
+Source:    $(PHOBOSSRC std/_experimental/_ndslice/_algorithm.d)
+
+Macros:
+SUBREF = $(REF_ALTTEXT $(TT $2), $2, std,experimental, ndslice, $1)$(NBSP)
+T2=$(TR $(TDNW $(LREF $1)) $(TD $+))
+T6=$(TR $(TDNW $(LREF $1)) $(TD $2) $(TD $3) $(TD $4) $(TD $5) $(TD $6))
+*/
+module std.experimental.ndslice.algorithm;
+
+import std.traits;
+import std.meta;
+
+import std.experimental.ndslice.internal;
+import std.experimental.ndslice.slice;
+
+private template TensorFronts(size_t length)
+{
+    static if (length)
+    {
+        enum i = length - 1;
+        enum TensorFronts = TensorFronts!(length - 1) ~ "tensors[" ~ i.stringof ~ "].front, ";
+    }
+    else
+    {
+        enum TensorFronts = "";
+    }
+}
+
+private void checkShapesMatch(bool seed, Args...)(auto ref Args tensors)
+{
+    enum msg = seed ?
+        "all arguments except the first (seed or needle) must be tensors" :
+        "all arguments must be tensors"
+        ~ tailErrorMessage!();
+    enum msgShape = "all tensors must have the same shape"  ~ tailErrorMessage!();
+    foreach (i, Arg; Args)
+    {
+        static assert (is(Arg == Slice!(N, Range), size_t N, Range), msg);
+        static if (i)
+        {
+            static assert (tensors[i].N == tensors[0].N, msgShape);
+            assert(tensors[i].shape == tensors[0].shape, msgShape);
+        }
+    }
+}
+
+private template naryFun(bool hasSeed, size_t argCount, alias fun)
+{
+    static if (argCount + hasSeed == 1)
+    {
+        import std.functional : unaryFun;
+        alias naryFun = unaryFun!fun;
+    }
+    else
+    static if (argCount + hasSeed == 2)
+    {
+        import std.functional : binaryFun;
+        alias naryFun = binaryFun!fun;
+    }
+    else
+    {
+        alias naryFun = fun;
+    }
+}
+
+/++
+Implements the homonym function (also known as `transform`) present
+in many languages of functional flavor. The call `ndMap!(fun)(tensor)`
+returns a tensor of which elements are obtained by applying `fun`
+for all elements in `tensor`. The original tensors are
+not changed. Evaluation is done lazily.
+
+Note:
+    $(SUBREF iteration, transposed) and
+    $(SUBREF selection, pack) can be used to specify dimensions.
+Params:
+    fun = One or more functions.
+    tensor = An input tensor.
+Returns:
+    a tensor with each fun applied to all the elements. If there is more than one
+    fun, the element type will be `Tuple` containing one element for each fun.
+See_Also:
+    $(REF map, std,algorithm,iteration)
+    $(HTTP en.wikipedia.org/wiki/Map_(higher-order_function), Map (higher-order function))
++/
+template ndMap(fun...)
+    if (fun.length)
+{
+    ///
```

Comment:
Hmm we will switch to ddox with the upcoming 2.072 release, so that won't work :/